packages/langium/src/parser/langium-parser.ts

```diff
@@ -232,15 +233,16 @@ export class LangiumParser extends AbstractLangiumParser {
 
     parse<T extends AstNode = AstNode>(input: string, options: ParserOptions = {}): ParseResult<T> {
         this.nodeBuilder.buildRootNode(input);
-        const lexerResult = this.lexer.tokenize(input);
+        const lexerResult = this.lexerResult = this.lexer.tokenize(input);
         this.wrapper.input = lexerResult.tokens;
         const ruleMethod = options.rule ? this.allRules.get(options.rule) : this.mainRule;
         if (!ruleMethod) {
             throw new Error(options.rule ? `No rule found with name '${options.rule}'` : 'No main rule available.');
         }
         const result = ruleMethod.call(this.wrapper, {});
-        this.nodeBuilder.addHiddenTokens(lexerResult.hidden);
+        this.appendHiddenTokens(lexerResult.hidden);
```

Comment:
Thinking loud, I assume that these hidden tokens are sorted by position. This can be used to optimize their insertion (correct me if I am wrong, in my eyes we can be faster, or? But maybe I missed a detail that does this already?).

Having a `CstNode traverser` could be a solution. Like:

```ts
interface CstNodeTraverser {
  stack: [CompositeCstNode, ChildIndex][];
  advanceBefore(token): void;
  insert(token): void
}
```

The `advanceBefore` method traverses the tree from its last position, not from the root.
I saw that the node builder has already a stack. But I assume that everything is completed and the stack is empty, before you start to add hidden nodes.

Reply:
Not necessary to make it this complicated, but I was able to get further performance improvements out of this by adding multiple tokens at the same time. Thanks.